v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/bigtablechangestreamstopubsub/FailsafeModJsonToPubsubMessageTransformer.java

```diff
@@ -0,0 +1,210 @@
+                  .withOutputTags(transformOut, TupleTagList.of(transformDeadLetterOut)));
+      out.get(transformDeadLetterOut).setCoder(failsafeModJsonToPubsubMessageOptions.getCoder());
+      return out;
+    }
+
+    /**
+     * The {@link FailsafeModJsonToPubsubMessageFn} converts a JSON string wrapped in {@link
+     * FailsafeElement} to a {@link PubsubMessage}.
+     */
+    public static class FailsafeModJsonToPubsubMessageFn
+        extends DoFn<FailsafeElement<String, String>, PubsubMessage> {
+      private final PubSubUtils pubSubUtils;
+      public TupleTag<PubsubMessage> transformOut;
+      public TupleTag<FailsafeElement<String, String>> transformDeadLetterOut;
+      private transient Publisher publisher;
+
+      public FailsafeModJsonToPubsubMessageFn(
+          PubSubUtils pubSubUtils,
+          TupleTag<PubsubMessage> transformOut,
+          TupleTag<FailsafeElement<String, String>> transformDeadLetterOut) {
+        this.pubSubUtils = pubSubUtils;
+        this.transformOut = transformOut;
+        this.transformDeadLetterOut = transformDeadLetterOut;
+      }
+
+      @Setup
+      public void setUp() {
+        try {
+          final TopicName projectTopicName =
+              TopicName.of(
+                  pubSubUtils.getDestination().getPubSubProject(),
+                  pubSubUtils.getDestination().getPubSubTopicName());
+          publisher = Publisher.newBuilder(projectTopicName).build();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      @Teardown
+      public void tearDown() {
+        try {
+          if (publisher != null) {
+            publisher.shutdown();
+            publisher.awaitTermination(5, TimeUnit.MINUTES);
+          }
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) {
+        LOG.info("Reading a failsafeModJsonString");
+        FailsafeElement<String, String> failsafeModJsonString = context.element();
+
+        try {
+          PubsubMessage pubSubMessage =
+              publishModJsonStringAsPubsubMessage(failsafeModJsonString.getPayload());
+
+          context.output(pubSubMessage);
+        } catch (Exception e) {
+          LOG.error("Writing to transform dead letter out", e);
+          context.output(
+              transformDeadLetterOut,
+              FailsafeElement.of(failsafeModJsonString)
+                  .setErrorMessage(e.getMessage())
+                  .setStacktrace(Throwables.getStackTraceAsString(e)));
+        }
+      }
+
+      /* Schema Details:  */
+      private PubsubMessage publishModJsonStringAsPubsubMessage(String modJsonString)
+          throws Exception {
+        ObjectNode modObjectNode = (ObjectNode) new ObjectMapper().readTree(modJsonString);
+        String changeJsonString = Mod.fromJson(modObjectNode.toString()).getChangeJson();
```

Comment:
These two lines look odd. 

We are supposedly already have modJsonString as input. Then we parse it into object tree (modObjectNode) and then we serialize modObjectNode into a String (is it going to be JSON String?), but then we parse it again into Mod to extract .getChangeJson().

Reply:
Looks like you should be able to do just 

`String changeJsonString = Mod.fromJson(modJsonString).getChangeJson();`